File: cai/88.py
def merge_sort(num):
    if len(num) <= 1:
        return num
    l = len(num)
    n = l // 2
    m = l - n
    num1 = merge_sort(num[0:m])
    num2 = merge_sort(num[m:l])
    i, j = 0, 0
    for k in range(l):
        if num1[i] <= num2[j]:
            num[k] = num1[i]
            i += 1
        else:
            num[k] = num2[j]
            j += 1
        if (i == m) and (j < n):
            num[k + 1:] = num2[j:l]
            break
        elif (i < m) and (j == n):
            num[k + 1:] = num1[i:m]
            break
    return num


class Solution:
    def merge(self, nums1: list, m: int, nums2: list, n: int):
        """
        Do not return anything, modify nums1 in-place instead.
        """
#        n1 = merge_sort(nums1[0:m])
#        n2 = merge_sort(nums2)
        n1=nums1[0:m]
        n2=nums2
        if (m == 0) or (n == 0):
            nums1[:] = n1 + n2
        else:
            i, j = 0, 0
            l=m+n
            for k in range(l):
                if n1[i] <= n2[j]:
                    nums1[k] = n1[i]
                    i += 1
                else:
                    nums1[k] = n2[j]
                    j += 1
                if (i == m) and (j < n):
                    nums1[k + 1:] = n2[j:l]
                    break
                elif (i < m) and (j == n):
                    nums1[k + 1:] = n1[i:m]
                    break

File: xiong/88.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        if(m == 0){
            nums1 = nums2;
            return;
        }
        if(n == 0){
            return;
        }
        int start1 = 0, start2 = 0, max1 = nums1[m - 1], tl = m + n;
        for(start1 = 0; start1 < m + n; ++start1){
            if(nums1[start1] == max1 && nums1[start1] <= nums2[start2]) {
                for(; start2 < n; ++start2){
                    // cout<<start2<<endl;
                    nums1.insert(nums1.begin() + m, nums2[start2]);
                    m++;
                }
                break;
            }
            if(nums1[start1] < nums2[start2]) continue;
            

            // nums1[start1] >= nums[start2]
            nums1.insert(nums1.begin() + start1, nums2[start2]);
            m++;
            start2++;
            if(start2 >= n) break;
        }
        for(int i = nums1.size() - 1; i >= tl; --i){
            nums1.erase(nums1.begin() + i);
        }
    }
};